### PR TITLE
fix(gradle): Enhance parsing

### LINKED ISSUE #12518: Improve Gradle parser
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Describe the bug

There is some space for quick improvements for Gradle parser.
It should fix the concerns described in #12240 (see also minimal repro).

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Handle two-part variable definitions (e.g. `foo.bar = 'baz'`)
- Change processing order: `versions.gradle` is always first, `build.gradle` is always last of `*.gradle.files`
- Handle assignment of dep string (e.g. update dependencies defined as `foobar = 'foo:bar:1.2.3'`)
- Skip-reason for incorrectly parsed deps, e.g. `createXmlValueRemover("defaults", "integer", "integer")`

## Context:

Closes #12518

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Would be hard to test without #12533 being merged

**Comment by zharinov** [comment]
Tested with: https://github.com/renovate-testing/test-12518-gradle-parser-fixes/pulls

**Comment by zharinov** [comment]
See also: https://github.com/renovate-testing/test-12518-architecture-components-samples/pulls

File: lib/manager/gradle/shallow/utils.ts [hunk 1]
```diff
@@ -95,9 +95,19 @@ export function interpolateString(
   return resolvedSubstrings.join('');
 }
 
+export function isGradleVersionsFile(path: string): boolean {
+  const filename = upath.basename(path).toLowerCase();
+  return /^versions\.gradle(?:\.kts)?$/.test(filename);
```
Comment (viceice) [inline lib/manager/gradle/shallow/utils.ts:100]: - Use invariant case flag to regex, so no need for lower case.
- Should use re2.
- Regex should be module const.

What about this?
```suggestion
  return /(^|\/|\\)versions\.gradle(?:\.kts)?$/i.test(path);
```

File: lib/manager/gradle/shallow/utils.ts [hunk 2]
```diff
@@ -95,9 +95,19 @@ export function interpolateString(
...
+}
+
+export function isGradleBuildFile(path: string): boolean {
+  const filename = upath.basename(path).toLowerCase();
+  return /^build\.gradle(?:\.kts)?$/.test(filename);
```
Comment (viceice) [inline lib/manager/gradle/shallow/utils.ts:105]: See above

File: lib/manager/gradle/shallow/utils.ts [hunk 3]
```diff
@@ -95,9 +95,23 @@ export function interpolateString(
   return resolvedSubstrings.join('');
 }
 
+const gradleVersionsFileRegex = regEx('^versions\\.gradle(?:\\.kts)?$', 'i');
+const gradleBuildFileRegex = regEx('^build\\.gradle(?:\\.kts)?$', 'i');
+const gradleFileRegex = regEx('\\.gradle(?:\\.kts)?$', 'i');
+
+export function isGradleVersionsFile(path: string): boolean {
+  const filename = upath.basename(path);
```
Comment (viceice) [inline lib/manager/gradle/shallow/utils.ts:103]: I still think we could do this with Regex only, but I'm ok with current way.

**Review by viceice — APPROVED** [review]

**Pull request merged**